nemo/collections/nlp/models/question_answering/qa_gpt_model.py

```diff
@@ -0,0 +1,485 @@
+        )
+        input_ids, input_attn_mask = self._transfer_tensors_to_device(
+            [encoded_dict["input_ids"], encoded_dict["attention_mask"]], device
+        )
+        labels = GPTQADataset.get_labels_from_inputs(
+            input_ids, training_mask_end, input_attn_mask, self.tokenizer.tokenizer,
+        )
+        if len(labels.shape) == 1:
+            labels = torch.unsqueeze(labels, 0)
+        labels = labels.to(device)
+
+        return input_ids, input_attn_mask, labels, generated_texts
+
+    def _generate_candidates(self, input_ids, input_attn_mask, training_mask_end):
+        num_tokens_to_generate = self.cfg.tokens_to_generate
+        if self.cfg.library == "huggingface":
+            generated_token_ids = []
+            max_length = 0
+            for i in range(input_ids.size(0)):
+                param_dict = {
+                    "input_ids": input_ids[i : i + 1, : training_mask_end[i]],
+                    "attention_masks": input_attn_mask[i : i + 1, : training_mask_end[i]],
+                    "max_length": training_mask_end[i] + num_tokens_to_generate,
+                    "pad_token_id": self.tokenizer.tokenizer.pad_token_id,
+                }
+                generated_token_ids.append(self.language_model.generate(**param_dict, skip_special_tokens=True))
+                max_length = max(max_length, generated_token_ids[-1].size(1))
+
+            # pad each generated to ensure they are of same length in dim 1, therefore stack-able
+            generated_token_ids = [
+                torch.cat(
+                    [i, torch.ones((1, max_length - i.size(1))).to(i.device) * self.tokenizer.tokenizer.pad_token_id],
+                    axis=-1,
+                )
+                for i in generated_token_ids
+            ]
+            generated_token_ids = torch.cat(generated_token_ids, axis=0)
+            generated_answers = self._get_answers_from_generated_tokens(
+                generated_token_ids, training_mask_end=training_mask_end
+            )
+
+        elif self.cfg.library == 'megatron':
+            raise ValueError("Megatron Generation is not supported by GPTQAModel")
+
+        return generated_answers
+
+    def _get_answers_from_generated_tokens(self, token_ids, training_mask_end=None):
+        answers = []
+        for i in range(token_ids.size(0)):
+            start_point = 0 if training_mask_end is None else training_mask_end[i].item()
+            stop_point = token_ids.size(1)
+
+            for j in range(start_point, stop_point):
+                if token_ids.data[i, j] == self.tokenizer.tokenizer.pad_token_id:
+                    stop_point = j
+                    break
+
+            curr_answer = self.tokenizer.tokenizer.decode(
+                token_ids[i, start_point:stop_point], skip_special_tokens=True
+            ).strip()
+            answers.append(curr_answer)
+
+        return answers
+
+    def _get_raw_scores(self, examples, preds):
```

Comment:
This seems exactly like in BERT, maybe we can have a base QA model that does all these? Alternatively this method can be delegate to the QA metrics class

Reply:
For instance, the metrics class can have a method called `_get_raw_scores(self, examples, preds):` that various models call

Reply:
> This seems exactly like in BERT, maybe we can have a base QA model that does all these? Alternatively this method can be delegate to the QA metrics class

Would we need to extend metrics for generative QA when we get to free form answering like Context + Question -> Fluent response?

Reply:
I think the metrics in this case will remain as F1 and Exact Match? I've also used BLEU before but iirc, they highly correlate with F1 so I think it will be fine to keep using F1 and Exact Match (also easier to understand). BLEU is most helpful when there are multiple acceptable answers, each of which is slightly different from one another (i.e. in Machine Translation).